ENH: Native conversion from/to scipy.sparse matrix to SparseDataFrame

kernc opened the pull request:
- [x] closes #4343
 - [x] tests added / passed
 - [X] passes ``git diff upstream/master | flake8 --diff``
 - [X] whatsnew entry

kernc committed:
ENH: Native conversion from/to scipy.sparse matrix to SparseDataFrame

kernc commented:
The tests require scipy, however scipy is not a pandas dependency. What do?

jreback commented:
scipy is already included in many of the test runs e.g. https://github.com/pandas-dev/pandas/blob/master/ci/requirements-2.7.run

you need to use tm._skip_if_no_scipy in your tests (assume u don't need a specific version required)

in code it's a bit more tricky - l will comment on review in a bit

jreback commented:
see for example .interpolate where we dispatch to scipy (or raise an error if it's not installed) for some methods

jreback commented:
so this API is a bit at odds with what was done in ``SparseSeries.from_coo``, meaning that we have a specific 'type' factory function. We don't actually follow this pattern anywhere else (e.g. even ``DataFrame.from_dict`` is just a small wrapper on the ``DataFrame`` constructor).

So for API consistency, I think it makes sense to change ``SparseSeries.from_coo`` in a similar way.

@wesm IIRC you had some thoughts w.r.t. to this.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -153,6 +153,28 @@ New Behavior:
+
+SciPy sparse matrix from/to SparseDataFrame
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Pandas now supports creating SparseDataFrames directly from ``scipy.sparse.spmatrix`` instances. E.g.
```
add the issue number after instanes. Remove the 'E.g'.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -153,6 +153,28 @@ New Behavior:
+   sdf = pd.DataFrame(sp_arr)
+   sdf
+
+To convert a SparseDataFrame back to scipy sparse matrix in COO format, you can use:
```
add this example to the sparse docs as well.

jreback commented on pandas/sparse/frame.py:
```diff
@@ -25,9 +25,14 @@
 from pandas.util.decorators import Appender
 import pandas.core.ops as ops
 
+try:
```
so we don't want to do an import here, which will affect all of pandas. 

- create a function ``pandas.types.common.is_scipy_sparse``

then can either: 

- do the import (and thus check ``isinstance``), you *may* need to jump some hoops to avoid repeatedly trying this import (you make a ``_try_import_scipy_sparse`` function), then assign the results to a module level variable on first use. see in ``pandas.io.pytables`` how ``_tables()`` works.

- OR, and maybe better, if scipy sparse matrixes are duck-type inferable, IOW, if its possible to check if they have a couple of key attributes then I would just do this.

jreback commented on pandas/sparse/frame.py:
```diff
@@ -85,24 +90,20 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
         elif isinstance(data, BlockManager):
             mgr = self._init_mgr(data, axes=dict(index=index, columns=columns),
                                  dtype=dtype, copy=copy)
+        elif isinstance(data, spmatrix):
```
so this will change to ``is_scipy_sparse_matrix``

jreback commented on pandas/sparse/frame.py:
```diff
@@ -175,6 +176,33 @@ def _init_dict(self, data, index, columns, dtype=None):
 
     def _init_matrix(self, data, index, columns, dtype=None):
         data = _prep_ndarray(data, copy=False)
```
can you add a 1-liner doc (what this function is doing)

jreback commented on pandas/sparse/frame.py:
```diff
@@ -175,6 +176,33 @@ def _init_dict(self, data, index, columns, dtype=None):
+        bindex = np.arange(N, dtype=np.int32)
+
+        sdict = {}
+        values = Series(data.data, index=data.row)
```
you can add ``copy=False``

kernc replied:
I did, but isn't `copy=False` default?

jreback commented on pandas/sparse/frame.py:
```diff
@@ -187,9 +215,84 @@ def _init_matrix(self, data, index, columns, dtype=None):
 
-        data = dict([(idx, data[:, i]) for i, idx in enumerate(columns)])
-        return self._init_dict(data, index, columns, dtype)
+    def as_matrix(self, columns=None, sparse=False):
```
I understand why you are changing / adding this. BUT this is actually a major API change. This is what ``.values`` calls and as a result the return types are: dense numpy arrays / pandas extension types (categorical).

Not against this (in fact I think its a good idea), but split this out into a separate issue. Then the impl would actually be done in the block manager.

jreback commented on pandas/sparse/frame.py:
```diff
@@ -187,9 +215,84 @@ def _init_matrix(self, data, index, columns, dtype=None):
+    def to_coo(self):
+        """
+        Convert the frame to its SciPy sparse COO matrix representation.
+
```
add a versionadded

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1118,31 @@ def test_isnotnull(self):
+    def test_from_to_scipy(self):
+        # GH 4343
+        try:
+            from scipy.sparse import csr_matrix
```
use ``tm._skip_if_no_scipy``

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1118,31 @@ def test_isnotnull(self):
+        for index, columns in ((list('abc'), list('def')),
+                               (None, None)):
+            sdf = pd.SparseDataFrame(spm, index=index, columns=columns)
+
```
I would rather you actually construct the SDF directly and simply use ``tm.assert_sp_frame_equal`` which already does all of this.

Further check for various fill types as well.

I would write this using ``pytest.parametrize`` (as our new way of doing things)

jreback commented on pandas/sparse/frame.py:
```diff
@@ -175,6 +176,33 @@ def _init_dict(self, data, index, columns, dtype=None):
+        return self._init_dict(data, index, columns, dtype)
+
+    def _init_spmatrix(self, data, index, columns, dtype=None):
+        index, columns = self._prep_index(data, index, columns)
```
1-liner here as well

jreback reviewed (changes requested):
looks generally good, but some pandas style requirements (on the code) and methodologies

wesm commented:
I don't have too many opinions on this feature

jorisvandenbossche commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -153,6 +153,28 @@ New Behavior:
+   arr[arr < .9] = 0
+   sp_arr = csr_matrix(arr)
+   sp_arr
+   sdf = pd.DataFrame(sp_arr)
```
Should this be `SparseDataFrame(sp_arr)` instead of DataFrame?

jorisvandenbossche commented:
Never use sparse so also not much opinion, but a question: there already exists a `SparseSeries.from_coo` method. Would it therefore make sense to have a `SparseDataFrame.from_csr` method instead of overloading the constructor?

jreback commented:
well @jorisvandenbossche that's exactly the question. just because we have ``.from_coo`` doesn't mean which *should* have specific constructors. In fact we *don't* for *anything* else. So IMHO we should just fold these into the constructor (IOW, leave this PR alone and deprecate and change ``.from_coo``

kernc commented on pandas/types/common.py:
```diff
@@ -59,6 +59,21 @@ def is_sparse(array):
+def is_scipy_sparse(array):
+    """ return if we are a scipy.sparse.spmatrix """
+    global _spmatrix
+    if _spmatrix is None:
```
Avoided importing in a separate function as this seemed concise enough.

Am not aware of any features that would unambiguously discern scipy sparse matrices. We'd have to check for what we use later on, namely `.shape` and `.tocoo(copy: bool) -> scipy.sparse.coo_matrix`. Checking the type directly looks cleaner.

jreback replied:
yeah ok that's fine

kernc commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1120,28 @@ def test_isnotnull(self):
                             'B': [True, False, True, True, False]})
         tm.assert_frame_equal(res.to_dense(), exp)
 
+    @pytest.mark.importorskip('scipy.sparse')
```
Used `tm._skip_if_no_scipy()` first, but then opted for this. Is that ok?

jreback replied:
yes that's fine if on that test

jreback replied:
actually on 2nd thought, use the ``tm._skip....`` as we are consistently using that. (does the same thing)

kernc commented on doc/source/sparse.rst:
```diff
@@ -186,9 +186,32 @@ the correct dense result.
+.. versionadded:: 0.20.0
 
-A :meth:`SparseSeries.to_coo` method is implemented for transforming a ``SparseSeries`` indexed by a ``MultiIndex`` to a ``scipy.sparse.coo_matrix``.
+Pandas supports creating sparse dataframes directly from ``scipy.sparse`` matrices.
```
Normally, versionadded parts are appended at the end, but I thought this prominent, hopefully non-experimental feature should be exposed at the top ...

jreback replied:
no that's fine, just mention, starting in 0.20.0 (or can use a versionadded tag)

kernc commented:
> So for API consistency, I think it makes sense to change SparseSeries.from_coo in a similar way.

What way is that? SparseSeries accepting 1×N sparse matrices in the constructor?\

> Would it make sense to have a `SparseDataFrame.from_csr` method instead of overloading the constructor?

There are [other types of sparse matrices](https://docs.scipy.org/doc/scipy-0.18.1/reference/sparse.html#sparse-matrix-classes) and while the user could be asked to convert to a specific format, why should they be forced to. Conversion from one sparse format to another is cheap (relative to any operation that makes anything dense). Several spmatrix methods do the conversion to the optimal type for the operation internally as needed. And spmatrix objects are still _de facto_ sparse structures in this stack, so accepting them in the constructor would not be an unexpected functionality.

jreback commented:
> There are other types of sparse matrices and while the user could be asked to convert to a specific format, why should they be forced to. Conversion from one sparse format to another is cheap (relative to any operation that makes anything dense). Several spmatrix methods do the conversion to the optimal type for the operation internally as needed. And spmatrix objects are still de facto sparse structures in this stack, so accepting them in the constructor would not be an unexpected functionality.

so the question is to we have several ``.from_type_of_sparse_conversion`` routines, OR attempt to figure this out in the constructor?

I think it should be consistent 1 way or the other. (e.g. NO implicit construction in the constructor / maybe raise an error actually, with ``.from_*`` methods), or ALL constructions in the constructor.

kernc commented:
There are no relevant `from_*` methods on SparseDataFrame currently, only one on SparseSeries.

As is, we already support _all_ scipy.sparse.spmatrix subclasses through the constructor, so individual per-type `from_*` methods are unnecessary. But I'm happy to add a thin factory method `SparseDataFrame.from_scipy` (or equivalent) that calls the constructor if the consensus is this importantly aids readability. (I'm actually :-1:. SciPy sparse matrices are as ubiquitous in this stack that accepting them in the SparseDataFrame constructor is as natural as accepting ndarrays in DataFrame's.)

jreback commented:
> As is, we already support all scipy.sparse.spmatrix subclasses through the constructor, so individual per-type from_* methods are unnecessary. But I'm happy to add a thin factory method SparseDataFrame.from_scipy (or equivalent) that calls the constructor if the consensus is this importantly aids readability. (I'm actually 👎. SciPy sparse matrices are as ubiquitous in this stack that accepting them in the SparseDataFrame constructor is as natural as accepting ndarrays in DataFrame's.)

@kernc so I think you answered my question. If we can unambiguously accept scipy sparse matricies w/o having to have a ``.from_*`` then I am all for it. So would then deprecate``SparseSeries.from_coo``, and move inside the constructor instead.

kernc commented:
But deprecate `SparseSeries.from_coo` in favor of what? Accepting a 1×N scipy matrix in the constructor and making it a Series, _and_ accepting a N×M scipy matrix and making it a Series indexed by a MultiIndex?

Basically, whatever but just accepting stuff in the constructor?

jreback commented:
> But deprecate SparseSeries.from_coo in favor of what? Accepting a 1×N scipy matrix in the constructor and making it a Series, and accepting a N×M scipy matrix and making it a Series indexed by a MultiIndex?

Simply moving the functionailty *into* the constuctor itself (almost exactly as it is).

IOW, something like


```
class SparseSeries(Series):

     ........

      if is_scipy_matrix(..):
           if is_coo_matrix(....):
                 # do what .from_coo does now
           else:
                raise NotImplentedError('on the todo list')
```

jreback commented:
@kernc of course you *may* still want these routines if you need to control the conversion process in some way; though we *maybe* could add those as optional params to the constructor themselves.

jreback commented on doc/source/sparse.rst:
```diff
@@ -186,9 +186,32 @@ the correct dense result.
+
+.. versionadded:: 0.16.0
+
+Additionally, an experimental :meth:`SparseSeries.to_coo` method is implemented for transforming a ``SparseSeries`` indexed by a ``MultiIndex`` to a ``scipy.sparse.coo_matrix``.
```
you can drop experimental

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -153,6 +153,28 @@ New Behavior:
+.. _whatsnew_0200.enhancements.scipy_sparse:
+
+SciPy sparse matrix from/to SparseDataFrame
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
add a ref to the new docs you created

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -153,6 +153,28 @@ New Behavior:
+SciPy sparse matrix from/to SparseDataFrame
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Pandas now supports creating sparse dataframes directly from ``scipy.sparse.spmatrix`` instances (:issue:`4343`).
+
```
mention if this *doesn't* work with particular scipy types (and in docs too)

jreback commented on pandas/sparse/frame.py:
```diff
@@ -174,7 +171,36 @@ def _init_dict(self, data, index, columns, dtype=None):
+    def _init_spmatrix(self, data, index, columns, dtype=None):
+        """ Init self from scipy.sparse matrix """
+        index, columns = self._prep_index(data, index, columns)
+        data = data.tocoo(copy=False)
```
ahh I see, tricky! does this have untoward side effects? (e.g. for certain matrices it is *bad*, meaning it doesn't work or forces a copy?)

kernc replied:
Works for any `scipy.sparse.spmatrix`. As per docs, it makes a copy if it needs to, but this is a copy of the sparse data, which is relatively cheap. Sparse matrix methods in SciPy do the conversion to the optimal type for the operation internally as needed, e.g. see `coo_matrix.maximum`, `lil_matrix.getcol`, `dok_matrix.__mul__` via `spmatrix._mul_sparse_matrix` ...

I am noting the transparent conversion (and potential copy) in the docs.

jreback replied:
ok that is fine

jreback commented on pandas/sparse/frame.py:
```diff
@@ -174,7 +171,36 @@ def _init_dict(self, data, index, columns, dtype=None):
+        data = data.tocoo(copy=False)
+        N = len(index)
+        bindex = np.arange(N, dtype=np.int32)
+
```
add a 1-liner on what you are doing

jreback commented on pandas/sparse/frame.py:
```diff
@@ -187,9 +213,47 @@ def _init_matrix(self, data, index, columns, dtype=None):
-        return self._init_dict(data, index, columns, dtype)
+    def to_coo(self):
+        """
+        Convert the frame to its SciPy sparse COO matrix representation.
```
does this require any special index? mention either way (and what happens to the index anyhow)? what about column names?

kernc replied:
There is no index (either top or side). A simple 2-D spmatrix (ndim==2) matrix is returned, laid out as the frame was. It's like `.values`, but for retaining sparsity.

What do I write?

jreback replied:
no, my point is that the index (of the DF) is discarded (well maybe the column names are kept?) Its find, just need to point that out.

kernc replied:
SciPy matrices are indexed (rows and columns) sequentially 0..(N-1), 0..(M-1), like ndarrays. Obviously, none of the DF indices are retained on the coo_matrix, only the raw values in the same shape. I don't know what to write. :blush: 

Perhaps I should change "Convert the frame ..." to "Return a SciPy sparse matrix ..." as nothing happens in-place.

jreback replied:
How does this compare to ``SparseSeries.to_coo`` then?

```
In [6]: >>> from numpy import nan
   ...: >>> s = Series([3.0, nan, 1.0, 3.0, nan, nan])
   ...: >>> s.index = MultiIndex.from_tuples([(1, 2, 'a', 0),
   ...:                                       (1, 2, 'a', 1),
   ...:                                       (1, 1, 'b', 0),
   ...:                                       (1, 1, 'b', 1),
   ...:                                       (2, 1, 'b', 0),
   ...:                                       (2, 1, 'b', 1)],
   ...:                                       names=['A', 'B', 'C', 'D'])
   ...: >>> ss = s.to_sparse()
   ...: 

In [7]: ss
Out[7]: 
A  B  C  D
1  2  a  0    3.0
         1    NaN
   1  b  0    1.0
         1    3.0
2  1  b  0    NaN
         1    NaN
dtype: float64
BlockIndex
Block locations: array([0, 2], dtype=int32)
Block lengths: array([1, 2], dtype=int32)

In [8]: ss.to_coo(row_levels=['A', 'B'],column_levels=['C', 'D'],sort_labels=True)
Out[8]: 
(<3x4 sparse matrix of type '<class 'numpy.float64'>'
        with 3 stored elements in COOrdinate format>,
 [(1, 1), (1, 2), (2, 1)],
 [('a', 0), ('a', 1), ('b', 0), ('b', 1)])

```

kernc replied:
Oh, at first glance, it's much simpler!

In your case:
```py
>>> ss.to_coo()
---------------------------------------------------------------------------
ValueError: Is not a partition because union is not the whole.

>>> # huh ???
```
With proposed (shown only for 2d case):
```py
>>> from scipy.sparse import coo_matrix
>>> sdf = pd.SparseDataFrame(coo_matrix([[0, 1, 0],
...                                      [1, 0, 0]]))

>>> sdf  # shit is a dataframe
     0    1   2
0  NaN  1.0 NaN
1  1.0  NaN NaN

>>> sdf[0]  # behaves like a dataframe
0    NaN
1    1.0
Name: 0, dtype: float64
BlockIndex
Block locations: array([1], dtype=int32)
Block lengths: array([1], dtype=int32)

>>> sdf.to_coo()  # no complex arguments
<2x3 sparse matrix of type '<class 'numpy.float64'>'
	with 2 stored elements in COOrdinate format>

>>> # gives back the kind of 2d matrix it was given
>>> # no COOrdinate (multi)index computing by hand
>>> sdf[[0, 1]].to_coo().toarray()  # just works
array([[0, 1],
       [1, 0]])
```
That was the idea, at least. :smiley:

jreback replied:
ok. btw. don't necessarily need to touch ``SparseSeries.to_coo`` (or as I mentioned before ``.from_coo`` in *this* PR. (of course its somewhat better / more consistent). More that we have a plan on *what* to do.

jreback commented on pandas/sparse/frame.py:
```diff
@@ -187,9 +213,47 @@ def _init_matrix(self, data, index, columns, dtype=None):
+
+        Notes
+        -----
+        The dtype will be the lowest-common-denominator type (implicit
```
mention that this is efficient (is it? does it copy? and if so what), mainly nice to have docs about this (and could refer main docs here as well).

kernc replied:
As efficient as `ndarray.astype()` and `np.concatenate()` are. Iterated on the number of columns as that was the easiest for me to understand. Doesn't make an unnecessary copy since `copy=False` was added just now. Thanks.

What do you mean by referring main docs?

jreback replied:
you can put a link in the doc-string itself to something in the main docs (a url). If you want to point to something more involved (this is an option).

kernc replied:
Dunno. I just copied these Notes from NDFrame.values, adapting a slight bit.

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1120,28 @@ def test_isnotnull(self):
                             'B': [True, False, True, True, False]})
         tm.assert_frame_equal(res.to_dense(), exp)
 
+    @pytest.mark.parameterize('index', [None, list('abc')])
```
nice!

kernc replied:
Not nice at all! Doesn't work! Without arguments' default values (`...=None`), all I'm getting is:

> TypeError: test_from_to_scipy() missing 3 required positional arguments: 'index', 'columns', and 'fill_value'

(With default values (`...=None`), those are the only values ever tested with, i.e. no param matrix.

https://paste.debian.net/917666/

Avoiding to read the whole of pytest docs, any clue as to where's the wrong I'm doing?

jreback replied:
no, what you need to do is create this as a function (and not put it in the class). parameterization doesn't work if:
``class Foo(tm.TestCase)`` which is how we start all of the classes. So easiest to make a stand alone function. You can change it to ``class Foo(object)``, but then all of the ``self.assert*`` methods have to be changed to ``pytest`` style. 

I have done that in a few cases.

kernc replied:
Right, got it. I guess it can't be a method on a `unittest.TestCase` instance.

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1120,28 @@ def test_isnotnull(self):
+    def test_from_to_scipy(self, index=None, columns=None, fill_value=None):
+        # GH 4343
+        tm._skip_if_no_scipy()
+        from scipy.sparse import csr_matrix
```
can you also make a test with exercises *all* of the scipy input matrix types? (you may want to make these as fixtures for convenience)

jreback commented on pandas/types/common.py:
```diff
@@ -59,6 +59,21 @@ def is_sparse(array):
     return isinstance(array, (ABCSparseArray, ABCSparseSeries))
 
 
+# oh the troubles to reduce import time
```
put at the top (the _spmatrix / comment)

jreback commented on pandas/types/common.py:
```diff
@@ -59,6 +59,21 @@ def is_sparse(array):
+    if _spmatrix is None:
+        try:
+            from scipy.sparse import spmatrix as _spmatrix
+        except ImportError:
```
can you add a test for this (pandas.types.tests.infererence) i think.

jreback reviewed (changes requested):
looks pretty good. few comments.

jreback commented:
you'll want to rebase on master as travis was having some issues

kernc commented:
Please give it another look. I made two major changes since the last iteration, namely:
* checking (testing) that dtypes are preserved when possible,
* the `default_fill_value` when unspecified set to `0` when initialized with a `spmatrix` as this is the implicit missing value in scipy sparse matrices.

kernc commented on pandas/sparse/array.py:
```diff
@@ -769,14 +769,17 @@ def make_sparse(arr, kind='block', fill_value=None):
+        mask = np.not_equal(arr, fill_value)
+        # In NumPy 1.12.0, not implemented at least for arr.dtype=str
+        if mask is NotImplemented:
+            mask = np.not_equal(arr.astype(object), fill_value)
```
I needed to change this bit in order to support adding a `str` (`object`) column to a `SparseDataFrame` with `default_fill_value=0`. See here: https://github.com/pandas-dev/pandas/pull/15497/files#diff-b30cded911d05ab8f4b2ab86c7dcab58R1169
Otherwise it would crash with some uninformative and unexpected error.

jreback replied:
*don't* use ``np.not_equal . what are you trying to do here?

jreback replied:
FYI #15587 is going to validate fill values (but of course let's fix here for what you need).

``isinstance(fill_value, compat.string_types)`` is idiom for checking string-likes (if its a scalar)

kernc replied:
Using `np.not_equal()` as an alternative to `!=`. Operator `!=` returns `False` comparing `arr` when `dtype=str` and `0`, issuing a FutureWarning that eventually, comparison will be element-wise. `np.not_equal()` doesn't issue such a warning but returns `NotImplemented`. One works with what one has. :smiley: 

So if I just want element-wise comparison every time, regardless of dtype, I'd better do:
```py
if pd.types.common.is_string_dtype(arr):
   arr = arr.astype(object)

mask = arr != fill_value
...
```
?

jreback replied:
no we *don't* use ``not_equal`` at all it in fact doesn't work on object arrays (well maybe in 1.12 it does, but its not compat).

use ``pandas.types.missing.array_equivalent`` if you really need, though ``isnull`` isself already calls this.

again what are you trying to do compare an array w/o a scalar?

kernc replied:
Well, `mask = arr != fill_value` test was there before since 2011. And this test fails when `arr.dtype.kind == 'U' and type(fill_value) != str` (i.e. comparing a vector of fixed-length strings to a non-string). Or at least `mask` gets set to _scalar_ `False` instead of a vector of boolean values, which then fails a few lines later on. It breaks on [this test line](https://github.com/pandas-dev/pandas/pull/15497/files#diff-b30cded911d05ab8f4b2ab86c7dcab58R1169), which, in my opinon, should generally pass.

jreback replied:
yeah the setting to scalar ``False``is the problem here. you'd expect an array, but I think numpy is broken for this (we had a similar issue in groupby.filter). in any event just do what you need to pass and I'll take a look.

kernc replied:
Right, continued in https://github.com/pandas-dev/pandas/pull/15497/files#r104474889.

kernc commented on pandas/sparse/array.py:
```diff
@@ -769,14 +770,20 @@ def make_sparse(arr, kind='block', fill_value=None):
+        # element-wise but just returns False if fill_value is not str,
+        # so cast to object comparison to be safe
+        if is_string_dtype(arr):
+            arr = arr.astype(object)
```
Perhaps the conversion should happen sooner, but certainly no later. Given the dtype-preserving behavior in SparseArray constructor, I doubt any sooner either.

jreback commented on pandas/types/common.py:
```diff
@@ -59,6 +62,17 @@ def is_sparse(array):
 
 
+def is_scipy_sparse(array):
+    """ return if we are a scipy.sparse.spmatrix """
```
this will keep trying to import scipy each time that it is called if scipy can *not* be imported. You have to set ``_is_scipy_sparse`` to *something* (rather than ``None``) if its ``None`` to start, so that it is *always* set here.

easiest would be to do

``_is_scipy_sparse = lambda x: False`` so this will just work.

jreback replied:
The other thing is, shouldn't this be something like

```
if _is_scipy_sparse is None:

    try:
        from scipy.sparse import issparse
    except ImportError:
        issparse = lambda x: False
    _is_scipy_sparse = issparse

return _is_scipy_sparse(array)
```

kernc replied:
Found it here: https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Import_Statement_Overhead Was worried about the same thing, but have tested in Py2 and Py3 that the global is indeed set correctly.

kernc replied:
Try:
```py
>>> def f():
...     global aaa
...     import sys as aaa

>>> f()
>>> aaa
<module 'sys' ...>
```

jreback replied:
didn't know that
it makes sense but is non obvious

jreback commented on pandas/sparse/frame.py:
```diff
@@ -77,29 +79,26 @@ def __init__(self, data=None, index=None, columns=None, default_kind=None,
 
         if default_fill_value is None:
-            default_fill_value = np.nan
+            default_fill_value = 0 if is_scipy_sparse(data) else np.nan
```
why is this?

kernc replied:
Well, SciPy sparse structures imply missing values are zeros. It felt like a sane default, but really it's better to keep the default _nan_. At least that way the user can decide, what fill values they want. (And I don't know how to "unfill" a sparse frame that has been `fillna`-d.)

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1120,58 @@ def test_isnotnull(self):
+def spmatrix(request):
+    tm._skip_if_no_scipy()
+    from scipy import sparse
+    yield getattr(sparse, request.param + '_matrix')
```
you don't use ``yield`` here, just return the result. If this *were* a yield fixture, then you would use ``@pytest.yield_fixture``, but no reason to use one as you don't need to do any cleanup actions.

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1120,58 @@ def test_isnotnull(self):
+@pytest.mark.parametrize('index', [None, list('abc')])
+@pytest.mark.parametrize('columns', [None, list('def')])
+@pytest.mark.parametrize('fill_value', [None, 0, np.nan])
+def test_from_to_scipy(spmatrix, index, columns, fill_value):
```
I would parametrize over dtypes too (all ints, uints, supported floats, object, bool).

kernc replied:
For the record, this makes a total of **1176 tests**, running some 20s on a very fast machine.

jreback replied:
woosh! ok if you want to cut it down a bit ok (also ok to leave this). our test suite is something like 400s to run (on 2 processors) anyhow.

kernc replied:
Can trim to `[object, bool, int, float, np.uint8, np.float16]` (504 tests) and assume the rest work.

kernc replied:
On the other hand, it was quite a pain to account for everything! But I think it's the generic way I'm testing in now that avoids the problems I was encountering at first.

jreback replied:
so you can leave out ``float16`` we don't test that *anywhere* anyhow (and barely support it).

yeah the rest look fine (though your comparisons maybe slightly tricky as int is platform dependent).

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1120,58 @@ def test_isnotnull(self):
+    if isinstance(fill_value, float):
+        arr = arr.astype(float)
+        arr[arr == 0] = np.nan
+    # If fill_value is None, set it to 0 for ndarray as value 0 is also the
```
blank line

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1120,58 @@ def test_isnotnull(self):
+    tm.assert_frame_equal(sdf.to_dense(), res.to_dense())
+    tm.assert_numpy_array_equal(res.values, arr)
+
+    # Assert spmatrices equal (operator!= not supported in scipy 0.8.0)
```
what does this comment mean? that's a really old version

kernc replied:
Yet that version complains in CircleCI-run tests.

jreback replied:
what? we don't install anything lower than 0.11

jreback commented on pandas/tests/types/test_inference.py:
```diff
@@ -30,11 +30,14 @@
 from pandas.types.missing import isnull
 from pandas.util import testing as tm
 
+from pandas.tests.sparse.test_frame import spmatrix  # noqa: F401
```
maybe just put ``spmatrix`` in ``pandas/tests/sparse/common.py`` (I would use ``conftest.py`` but not sure if these are picked up correctly)

kernc replied:
Right, even with this change, lint complains spmatrix is imported but unused.

jreback commented on pandas/tests/types/test_inference.py:
```diff
@@ -946,6 +949,12 @@ def test_nan_to_nat_conversions():
         assert (s[8].value == np.datetime64('NaT').astype(np.int64))
 
 
+def test_is_scipy_sparse(spmatrix):  # noqa: F811
```
why do you need the noqa on these?

kernc replied:
Here, lint complains: F811 redefinition of unused 'spmatrix' from line 39

jreback replied:
hmm yeah I guess it is 'removed' via the pytest mangling steps, odd. ok then.

kernc replied:
That's the thing with DSLs. Everyone needs to learn them, including linters. :smiley:

jreback commented on doc/source/sparse.rst:
```diff
@@ -186,9 +186,32 @@ the correct dense result.
+
+.. versionadded:: 0.16.0
+
+Additionally, a :meth:`SparseSeries.to_coo` method is implemented for transforming a ``SparseSeries`` indexed by a ``MultiIndex`` to a ``scipy.sparse.coo_matrix``.
```
can you show an example of ``SparseSeries.to_coo()`` (maybe just slice sdf)

kernc replied:
These lines are not an addition, Just slightly reworded from before. Usage examples abound below. I'd rather not make my own here as `series.to_coo()` currently doesn't work without arguments, which, to be honest, I'm having some trouble grasping.

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -189,6 +189,30 @@ You must enable this by setting the ``display.html.table_schema`` option to True
+
+SciPy sparse matrix from/to SparseDataFrame
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Pandas now supports creating sparse dataframes directly from ``scipy.sparse.spmatrix`` instances (:issue:`4343`). See the :ref:`documentation <sparse.scipysparse>` for more information.
```
can you put the issue number at the end of this paragraph (after information)

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -189,6 +189,30 @@ You must enable this by setting the ``display.html.table_schema`` option to True
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Pandas now supports creating sparse dataframes directly from ``scipy.sparse.spmatrix`` instances (:issue:`4343`). See the :ref:`documentation <sparse.scipysparse>` for more information.
+
+All sparse formats are supported, but matrices that aren't in *COOrdinate* format will be converted to it, copying the data as needed.
```
can you add a link directly to the sparse formats where *COOrdinate* is? (make that the text)

kernc replied:
Linked to sparse module where coo_matrix is like the fifth word vertically and COOrdinate is about approximately hcentered on the screen.

jreback commented on doc/source/sparse.rst:
```diff
@@ -186,9 +186,32 @@ the correct dense result.
+   sdf = pd.SparseDataFrame(sp_arr)
+   sdf
+
+All sparse formats are supported, but matrices that aren't in *COOrdinate* format will be converted to it, copying the data as needed. To convert a ``SparseDataFrame`` back to sparse SciPy matrix in COO format, you can use :meth:`SparseDataFrame.to_coo` method:
```
same link (see below) here.

jreback commented on pandas/sparse/frame.py:
```diff
@@ -25,6 +25,7 @@
                                    create_block_manager_from_arrays)
 import pandas.core.generic as generic
 from pandas.sparse.series import SparseSeries, SparseArray
+from pandas._sparse import BlockIndex, get_blocks
```
``from pandas.sparse.libsparse import BlockIndex, get_blocks`` (just changed)

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1126,62 @@ def test_isnotnull(self):
+                 dtype)
+    tm.assert_contains_all(sdf.dtypes, {np.dtype(res_dtype)})
+    tm.assert_equal(sdf.to_coo().dtype, res_dtype)
+
```
k this seems reasonable

jreback commented on pandas/tests/sparse/test_frame.py:
```diff
@@ -1118,6 +1126,62 @@ def test_isnotnull(self):
+    was_upcast = ((fill_value is None or is_float(fill_value)) and
+                  not is_object_dtype(dtype) and
+                  not is_float_dtype(dtype))
+    res_dtype = (bool if pd.types.common.is_bool_dtype(dtype) else
```
you can import this up top

jreback commented:
just some minor comments actually. lgtm. ping when all green and I think can merge. Will need to follow up to fix SparseSeries anyhow (to accept in the constructor) so can catch any fixes there.

kernc committed:
DOC: Fix some whatsnew/v0.20.0.txt sphinx warnings

kernc commented:
So this is green and kinda works. Feel free to edit at will.

jreback commented:
@kernc will have a look tomorrow and merge.

jreback commented:
@sinhrks comments?

jreback commented:
thanks @kernc great PR!

love for you to take #15634 !

jreback commented:
@kernc not a big deal but https://travis-ci.org/pandas-dev/pandas/jobs/209914478

```
.../home/travis/miniconda3/envs/pandas/lib/python3.5/site-packages/scipy/sparse/sputils.py:114: UserWarning: object dtype is not supported by sparse matrices
  warnings.warn("object dtype is not supported by sparse matrices")
......................................................................................................................................................................................................................................................................................................................................................................./home/travis/miniconda3/envs/pandas/lib/python3.5/site-packages/scipy/sparse/sputils.py:114: UserWarning: object dtype is not supported by sparse matrices
  warnings.warn("object dtype is not supported by sparse matrices")
```

jreback commented:
should prob skip that test then (or catch warning)

kernc commented:
What seems to not be supported are ufuncs and binops upon them. Right, will skip it in #15634 if it can wait till next week.

jreback commented:
np thanks

jorisvandenbossche commented:
@kernc didn't go through the full reviewing discussion here, but was it ever discussed what the default fill value should be when converting a scipy sparse matrix to SparseDataFrame? 

Because now, it actually "changes" your data due to the default fill value of NaN:

```
In [27]: a = scipy.sparse.coo_matrix([[0., 1.], [2., 0.]])

In [28]: a.toarray()
Out[28]: 
array([[ 0.,  1.],
       [ 2.,  0.]])

In [29]: df = pd.SparseDataFrame(a)

In [30]: df
Out[30]: 
     0    1
0  NaN  1.0
1  2.0  NaN

In [31]: np.array(df)
Out[31]: 
array([[ nan,   1.],
       [  2.,  nan]])
```
